src/Compilers/CSharp/Portable/Binder/SwitchBinder_BindPatternSwitch.cs

```diff
@@ -12,44 +12,179 @@
+            if (labelsMap != null)
+            {
+                SourceLabelSymbol label;
+                if (labelsMap.TryGetValue(key, out label))
+                {
+                    Debug.Assert(label != (object)null);
+                    return label;
+                }
+            }
+
+            return null;
+        }
+
+        // Dictionary for the switch case/default labels.
+        // Case labels with a non-null constant value are indexed on their ConstantValue.
+        // Default label(s) are indexed on a special DefaultKey object.
+        // Invalid case labels with null constant value are indexed on the labelName.
+        private Dictionary<object, SourceLabelSymbol> _lazySwitchLabelsMap;
+        private static readonly object s_defaultKey = new object();
+
+        private Dictionary<object, SourceLabelSymbol> LabelsByValue
+        {
+            get
+            {
+                if (_lazySwitchLabelsMap == null && this.Labels.Length > 0)
+                {
+                    _lazySwitchLabelsMap = BuildLabelsByValue(this.Labels);
+                }
+
+                return _lazySwitchLabelsMap;
+            }
+        }
+
+        private static Dictionary<object, SourceLabelSymbol> BuildLabelsByValue(ImmutableArray<LabelSymbol> labels)
+        {
+            Debug.Assert(labels.Length > 0);
+
+            var map = new Dictionary<object, SourceLabelSymbol>(labels.Length, new SwitchConstantValueHelper.SwitchLabelsComparer());
+            foreach (SourceLabelSymbol label in labels)
+            {
+                SyntaxKind labelKind = label.IdentifierNodeOrToken.Kind();
+
+                if (labelKind == SyntaxKind.CaseSwitchLabel ||
+                    labelKind == SyntaxKind.DefaultSwitchLabel)
+                {
+                    object key;
+                    var constantValue = label.SwitchCaseLabelConstant;
+                    if (constantValue != null)
+                    {
+                        // Case labels with a non-null constant value are indexed on their ConstantValue.
+                        key = constantValue;
+                    }
+                    else if (labelKind == SyntaxKind.DefaultSwitchLabel)
+                    {
+                        // Default label(s) are indexed on a special DefaultKey object.
+                        key = s_defaultKey;
+                    }
+                    else
+                    {
+                        // Invalid case labels with null constant value are indexed on the labelName.
+                        key = label.Name;
+                    }
+
+                    SourceLabelSymbol existingLabel;
+                    if (!map.TryGetValue(key, out existingLabel))
+                    {
+                        map.Add(key, label);
+                    }
+                    else
+                    {
+                        // If there is a duplicate label, ignore it. It will be reported in semantic analysis.
```

Comment:
Not sure what "semantic analysis" refers to. Aren't we are doing "semantic analysis" here?